Notifications

JssDWt opened the pull request:
Add notifications when the peer is offline.

- grpc server for registering a notification url for your pubkey.
- Sending a notification if the peer is offline and a registration was present.
- Hold a htlc a (configurable) little while when a notification was sent, waiting for the peer to come online.

This is a very basic implementation of https://github.com/breez/lspd/issues/63. It does not include the following features yet:

- Removing a webhook subscription after expiry or because it's not functioning or because the client wants to removee the webhook subscription.
- Log of published webhooks
- Retry mechanism for temporary failure to deliver the webhook call.
- Initial subscribed notification
- Chain notifications
- Solid mechanisms for not publishing duplicate notifications after restarts or replay of htlcs.

Note that the time period for holding the htlc might get tricky when you get close to the cltv delta. So the holding period should generally be short.

This PR replaces https://github.com/breez/lspd/pull/76

roeierez commented on interceptor/intercept.go:
```diff
@@ -85,66 +89,179 @@ func (i *Interceptor) Intercept(nextHop string, reqPaymentHash []byte, reqOutgoi
+					// So if this payment was not registered, this is a channel
+					// partner and we have to wait for the channel to become active
+					// before we can forward.
+					if !isRegistered {
```
In general I think we can see that if we don't have a registration in place we can wait for channel to be active from the beginning and only if we have registration we will wait for the peer to come online.
This could be also for a mobile user, not necessary a partner.

JssDWt replied:
- with channel partner I mean anyone we have a channel with (could be a mobile user. It's actually probably a mobile user).
- I do think we want to wait for 'normal forwards' right? Not just forwards where we need to open a channel.
- In general I think we only want to wait for the channel to become active if we've sent a notification. In any other case we want to fail fast.

roeierez commented on interceptor/intercept.go:
```diff
@@ -85,66 +89,179 @@ func (i *Interceptor) Intercept(nextHop string, reqPaymentHash []byte, reqOutgoi
+							}, nil
+						}
+					}
+				} else if isProbe {
```
The following two cases (else branches)  seems unrelated to the notification itself. Perhaps it is better to handle them after the notification code block (that start at` if (!isConnected) `)?
It seems that it will also allow extracting out the notification specific logic to its own function making the intercept logic easier to follow.

JssDWt replied:
The probe case is the hardest part of this flow. Maybe I need to know exactly when we want to return INCORRECT_OR_UNKNOWN_PAYMENT_DETAILS in case the payment hash is a probe payment hash.
Does this make sense?
- Payment is registered:
  - Peer is offline: Yes 
  - Peer is online:
    - Channel has not been opened: Yes
    - Channel has been opened: No

And when do we notify the client to come online when there is a probe payment hash? Never?

roeierez commented on cln/cln_client.go:
```diff
@@ -65,7 +66,7 @@ func (c *ClnClient) IsConnected(destination []byte) (bool, error) {
 
 	for _, peer := range peers {
-		if pubKey == peer.Id {
+		if pubKey == peer.Id && peer.Connected {
```
Not related as it is not introduced in this PR so just asking why don't we populate the peerID to the ListPeers endpoint? https://lightning.readthedocs.io/lightning-listpeers.7.html

JssDWt replied:
Good question. I don't think there's a good reason. Probably because glightning returns an error if the peer is not found.

JssDWt replied:
Fixed

roeierez reviewed (commented):
Looks good. Dropped two comments.

JssDWt commented:
@roeierez I added a commit to reduce nesting in the interceptor and hopefully make the flow a little easier to understand and fixed the ListPeers call by using GetPeer instead.

 One open question remaining here https://github.com/breez/lspd/pull/88#discussion_r1239541398
